File: Tutorial4-6SubComponentBuildersMultipleModulesAndActvities/src/main/java/com/example/tutorial4_6subcomponentbuildersmultiplemodulesandactvities/MainActivity.java
package com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities;

import android.content.Intent;
import android.content.SharedPreferences;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities.di.ApplicationComponent;
import com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities.di.DummyDependencyComponent;
import com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities.model.DummyDependency;
import com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities.model.SensorController;

import javax.inject.Inject;

/*
 ONLY one component can inject to an Object
 */
public class MainActivity extends AppCompatActivity {

    // Injected from ApplicationModule
    @Inject
    SharedPreferences sharedPreferences;

    // Injected from DummyDependencyModule
    @Inject
    DummyDependency dummyDependency;

    // Constructor Injection
    @Inject
    SensorController sensorController;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ApplicationComponent applicationComponent = ((MyApplication) getApplication()).getApplicationComponent();


        // Dummy Dependency
        DummyDependencyComponent dummyDependencyComponent = applicationComponent
                .dummyDependencyBuilder()
                .context(this)
                .build();

        dummyDependencyComponent.inject(this);


        Button button = findViewById(R.id.button);

        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(MainActivity.this, SecondActivity.class);
                startActivity(intent);
            }
        });

        TextView textView = findViewById(R.id.text_view);
        textView.setText("Dummy: " + dummyDependency.getApplicationName());

        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        ft.replace(R.id.container, new MyFragment()).commit();

        Toast.makeText(this, "MainActivity: " + sharedPreferences, Toast.LENGTH_SHORT).show();

    }

    @Override
    protected void onResume() {
        super.onResume();
        sensorController.onResume(SensorManager.SENSOR_DELAY_UI);
    }

    @Override
    protected void onPause() {
        super.onPause();
        sensorController.onPause();
    }
}

File: Tutorial4-6SubComponentBuildersMultipleModulesAndActvities/src/main/java/com/example/tutorial4_6subcomponentbuildersmultiplemodulesandactvities/MainActivity.kt
package com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities

import android.content.Intent
import android.content.SharedPreferences
import android.hardware.SensorManager
import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities.model.DummyDependency
import com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities.model.SensorController
import javax.inject.Inject

/*
 ONLY one component can inject to an Object
 */
class MainActivity : AppCompatActivity() {

    // Injected from ApplicationModule
    @Inject
   lateinit var sharedPreferences: SharedPreferences

    // Injected from DummyDependencyModule
    @Inject
    lateinit var dummyDependency: DummyDependency

    // Constructor Injection
    @Inject
    lateinit var sensorController: SensorController

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initInjection()

        findViewById<Button>(R.id.button).setOnClickListener {
            val intent = Intent(this@MainActivity, SecondActivity::class.java)
            startActivity(intent)
        }

        findViewById<TextView>(R.id.text_view).text = "Dummy:  ${dummyDependency.applicationName}"

        val ft = supportFragmentManager.beginTransaction()
        ft.replace(R.id.container, MyFragment()).commit()

        println("MainActivity sharedPreferences: $sharedPreferences")
        Toast.makeText(this, "MainActivity: $sharedPreferences", Toast.LENGTH_SHORT).show()
    }

    private fun initInjection() {
        val applicationComponent = (application as MyApplication).applicationComponent

        // Dummy Dependency
        val dummyDependencyComponent = applicationComponent
                .dummyDependencyBuilder()
                .context(this)
                .build()

        dummyDependencyComponent.inject(this)
    }

    override fun onResume() {
        super.onResume()
        sensorController.onResume(SensorManager.SENSOR_DELAY_UI)
    }

    override fun onPause() {
        super.onPause()
        sensorController.onPause()
    }
}